app/services/visit/integrity_checker.rb

```diff
@@ -0,0 +1,56 @@
+class Visit
+  module IntegrityChecker
+    def ensure_visit_integrity
+      return if required_information?
+      log_any_missing_information
+      flash[:notice] = I18n.t(
+        :ensure_visit_integrity,
+        scope: 'controllers.shared'
+      )
+      redirect_to edit_prisoner_details_path
+    end
+
+    def required_information?
+      return false unless visit
+      return false unless visit.prisoner?
+      return false unless visit.prison_name?
+      true
+    end
+
+    def log_any_missing_information
+      prefix = "Visit #{visit.visit_id} is missing a"
+      log_missing_prisoner(prefix)
+      log_missing_prisoner_number(prefix)
+      log_missing_prison_name(prefix)
+      log_missing_prison(prefix)
+      log_missing_prison_slots(prefix)
+    end
+
+    private
+
+    def log_missing_prisoner(prefix)
+      return if visit.prisoner?
+      Rails.logger.info("#{prefix} prisoner")
+    end
+
+    def log_missing_prisoner_number(prefix)
+      return if !visit.prisoner? || visit.prisoner_number?
+      Rails.logger.info("#{prefix} prisoner number")
```

Comment:
As above, I think the inversion of logic here (and in all the siblings of this) is a bit confusing (but that could be because it's early and I'm pre-:coffee:). When does this log? When there's a prisoner but no prisoner number? It might be clearer to wrap the if around the logger, or to make it a conditional line?